File: #162_Find_Peak_Element/Solution.java
public class Solution {
    public int findPeakElement(int[] nums) {
    	if (nums.length == 1) return 0;
        int lo = 0; int hi = nums.length - 1;
        return findPeakElement(nums, lo, hi);
    }

    private int findPeakElement(int[] nums, int lo, int hi) {
    	int mid = (lo + hi) / 2;
    	int prev, next;
    	if (mid - 1 < 0) prev = Integer.MIN_VALUE;
    	else prev = nums[mid - 1];
    	if (mid + 1 >= nums.length) next = Integer.MIN_VALUE;
    	else next = nums[mid + 1];
    	if (nums[mid] > prev && nums[mid] > next) return mid;
    	if (nums[mid] >= prev && nums[mid] < next) return findPeakElement(nums, mid + 1, hi);
    	if (nums[mid] < prev && nums[mid] > next) return findPeakElement(nums, lo, mid - 1);
    	return findPeakElement(nums, mid + 1, hi);
    }

    public static void main(String[] args) {
    	int[] s = {2, 1, 2};
    	Solution sol = new Solution();
    	System.out.println(sol.findPeakElement(s));
    }
}

File: #162_Find_Peak_Element/Solution.py
'''
O(log(n)) time
'''

class Solution:
    # @param nums, an integer[]
    # @return an integer
    def findPeakElement(self, nums):
        if not nums:
        	return -1
        return self.findPeak(nums, 0, len(nums) - 1)

    def findPeak(self, nums, lo, hi):
    	if lo == hi:
    		return lo
    	mid = lo + (hi - lo) / 2
    	if nums[mid] > nums[mid + 1]:
    		return self.findPeak(nums, lo, mid)
    	return self.findPeak(nums, mid + 1, hi)

sol = Solution()
print sol.findPeakElement([5, 4, 3, 2, 1])